astropy/nddata/tests/test_nddata.py

```diff
@@ -176,11 +179,11 @@ def test_nddata_init_data_maskedarray():
 @pytest.mark.parametrize("data", [np.array([1, 2, 3]), 5])
 def test_nddata_init_data_quantity(data):
     # Test an array and a scalar because a scalar Quantity does not always
-    # behaves the same way as an array.
+    # behave the same way as an array.
     quantity = data * u.adu
     ndd = NDData(quantity)
     assert ndd.unit == quantity.unit
-    assert_array_equal(ndd.data, np.array(quantity.value))
+    assert_array_equal(ndd.data, np.array(quantity))
```

Comment:
Is this an API change? Seems like it is different than the old behavior....

Reply:
Yes, and I want to ask your opinion on this more broadly. I would argue that it doesn't make sense to let `NDDataArray` store `Quantity` objects in the `data` attribute, since we would be duplicating the `unit` attribute in two places. Otherwise we'd have to put in a check to ensure that `nddata.data.unit == nddata.unit`. Similarly it would be confusing/inefficient if the mask of the `NDDataArray` was separate and/or copied from from `nddata.data.mask`.

I propose that it's clearer if `nddata.data` is _always_ an `np.ndarray`. The `NDDataArray` constructor should be able to find the unit and mask of the inputs, no matter how it is initialized. But the attributes stored on `NDDataArray` shouldn't be redundant.

Reply:
I'm going to need to think about this a bit. I don't think your argument is wrong -- currently we could have a mismatch between the unit of the `NDData` thing and the unit of the `NDData.data` thing, which seems error-prone.

The thing to think through is when to introduce a change like this, which I think depends in part on whose code would break if we changed this.

@saimn @DanRyanIrish @jehturner what do you think about this change?

Reply:
I don't understand this: `np.array(quantity)` and `np.array(quantity.value)` are identical: the first makes an array out of the quantity, the second makes an array of what already is an array of the values inside the quantity. Indeed, I think the clearest way to write that you want the `.data` not to have the unit (which I agree it should not have) is `assert_array_equal(ndd.data, quantity.value)`. Maybe even include `assert not hasattr(ndd.data, 'unit')` to make the intent clearer?

Reply:
Hmm, I'm not sure how this would affect us off hand (will try to check), but it sounds reasonable.

Reply:
What _does_ happen if you try to assign a `Quantity` to `ndd.data`? I agree with @bmorris3 that you probably don't want to _store_ it as one, but assigning one should probably do something sensible (copying the data while enforcing matching units?).

Reply:
It seems worth noting that `specutils` also stores its analagous `flux` attribute as a `Quantity` and requires it to be specified as `a` `Quantity` or `NDData`.

Reply:
If necessary, I can clobber a version of https://github.com/astropy/astropy-integration-testing/blob/main/tox.ini that would run with this PR (though given your fork doesn't have tags, it might also not work...). So, let me know if you want to try that. Thanks!

Reply:
> It seems worth noting that specutils also stores its analagous flux attribute as a Quantity and requires it to be specified as a Quantity or NDData.

@jehturner That's what the docstring used to say, but it only supports `Quantity`. See https://github.com/astropy/specutils/pull/1003.